auth_proxy/services.py

```diff
@@ -53,6 +54,25 @@ def audit(self, client_id):
         return self.db.session.query(Token).\
             filter_by(client_id=client_id)
 
+    def smart_token_credentials(self, grant_type, code=None, refresh_token=None):
+        """ Provide additional credentials required by a SMART token request.
+        """
+        if grant_type == 'authorization_code':
+            grant = self.db.session.query(Grant).\
+                filter_by(code=code).first()
+            user = grant.user
```

Comment:
This isn't exactly the right place for this check (I mean, logically -- technically it would work), but somewhere we should be ensuring that the `grant.user` is the same as the `request.user` at the time of the "token" call. That's an important defense against session fixation attacks.